File: algorithm/leetcodeCppTmp/plus-one.cpp
// 66. Plus One
// Difficulty: Easy
// Given a non-negative number represented as an array of digits, plus one to the number.

// The digits are stored such that the most significant digit is at the head of the list.

// Time:  O(n)
// Space: O(1)

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int carry = 1;
        for (auto it = digits.rbegin(); it != digits.rend(); ++it) {
            int sum = *it + carry;
            *it = sum % 10;
            carry = sum / 10;
        }
        if (carry == 1) {
            digits.insert(digits.begin(), carry);
        }
        return digits;  
    }
};

File: data-structure-and-algorithm/java/plus-one.java
/**
 * Plus One
 * 
 * Given a non-negative number represented as an array of digits, plus one to
 * the number.
 * 
 * The digits are stored such that the most significant digit is at the head of
 * the list.
 */

public class Solution {
    public int[] plusOne(int[] digits) {
        if (digits == null) {
            throw new NullPointerException();
        }

        for (int i = n - 1; i >= 0; --i) {
            if (digits[i] == 9) {
                digits[i] = 0; 
            } else {
                digits[i] += 1;
                return digits;
            }
        }

        if (digits[0] == 0) {
            int[] newDigits = new int[digits.length + 1];
            newDigits[0] = 1;
            return newDigits;
        }
        return digits;
    }
}
